Address Safer cpp failures in mac/ScrollerMac

iidmsa opened the pull request:
#### cfa4aa1a28ba4bf125f016ca98668f3bee093d28
<pre>
Address Safer cpp failures in mac/ScrollerMac
<a href="https://bugs.webkit.org/show_bug.cgi?id=291832">https://bugs.webkit.org/show_bug.cgi?id=291832</a>
<a href="https://rdar.apple.com/problem/149656606">rdar://problem/149656606</a>

Reviewed by Ryosuke Niwa.

Fix clang static analyzer warnings in mac/ScrollerMac.

* Source/WebCore/SaferCPPExpectations/NoUncountedMemberCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UnretainedCallArgsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UnretainedLocalVarsCheckerExpectations:
* Source/WebCore/page/scrolling/mac/ScrollerMac.h:
(WebCore::ScrollerMac::pair):
* Source/WebCore/page/scrolling/mac/ScrollerMac.mm:
(-[WebScrollbarPartAnimationMac setCurrentProgress:]):
(-[WebScrollerImpDelegateMac effectiveAppearanceForScrollerImp:]):
(-[WebScrollerImpDelegateMac setUpAlphaAnimation:featureToAnimate:animateAlphaTo:duration:]):
(WebCore::ScrollerMac::attach):
(WebCore::ScrollerMac::updateValues):
(WebCore::ScrollerMac::updateScrollbarStyle):
(WebCore::ScrollerMac::updatePairScrollerImps):
(WebCore::ScrollerMac::mouseEnteredScrollbar):
(WebCore::ScrollerMac::mouseExitedScrollbar):
(WebCore::ScrollerMac::lastKnownMousePositionInScrollbar const):
(WebCore::ScrollerMac::visibilityChanged):
(WebCore::ScrollerMac::updateMinimumKnobLength):
(WebCore::ScrollerMac::scrollbarState const):
* Source/WebCore/page/scrolling/mac/ScrollerPairMac.h:

Canonical link: <a href="https://commits.webkit.org/293949@main">https://commits.webkit.org/293949@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c27a6c03e27c41ec1dcf22a568926dbd4db821a7

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/100394 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/20046 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/10345 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/105530 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/50982 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/102435 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/20352 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/28520 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/76439 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/33495 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/103401 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/15587 "Found 1 new test failure: fast/forms/ios/file-upload-panel-dismiss-when-view-removed-from-window.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/90687 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/56794 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/15405 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/8688 "Passed tests") | [❌ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/50351 "Failed to compile WebKit") | 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/93056 "Build is in progress. Recent messages:") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/85319 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/8769 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/107885 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/27512 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/20184 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/85393 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/27875 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/86887 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/84931 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21596 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/29615 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/7346 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/21432 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/27447 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/32688 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/27258 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/30576 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/28816 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/page/scrolling/mac/ScrollerMac.mm:
```diff
@@ -100,18 +100,20 @@ - (void)setCurrentProgress:(NSAnimationProgress)progress
     else
         currentValue = progress;
 
+    RetainPtr<NSScrollerImp> scrollerImp = _scroller->scrollerImp();
```
Just say `RetainPtr` instead of `RetainPtr<NSScrollerImp>`.

iidmsa replied:
Fixed.

rniwa commented on Source/WebCore/page/scrolling/mac/ScrollerMac.mm:
```diff
@@ -204,7 +206,8 @@ - (NSAppearance *)effectiveAppearanceForScrollerImp:(NSScrollerImp *)scrollerImp
         return [NSAppearance currentDrawingAppearance];
     // The base system does not support dark Aqua, so we might get a null result.
-    if (auto *appearance = [NSAppearance appearanceNamed:_scroller->pair().useDarkAppearance() ? NSAppearanceNameDarkAqua : NSAppearanceNameAqua])
+    RetainPtr<NSString> appearanceName = _scroller->pair().useDarkAppearance() ? NSAppearanceNameDarkAqua : NSAppearanceNameAqua;
```
Oh, this is silly. We're fixing this in static analyzer: https://github.com/llvm/llvm-project/pull/136170
We should probably suppress this warning for now with `SUPPRESS_UNRETAINED_ARG`

iidmsa replied:
Done.

rniwa commented on Source/WebCore/page/scrolling/mac/ScrollerMac.mm:
```diff
@@ -217,9 +220,12 @@ - (void)setUpAlphaAnimation:(RetainPtr<WebScrollbarPartAnimationMac>&)scrollbarP
         scrollbarPartAnimation = nil;
     }
 
+    RetainPtr<NSScrollerImp> scrollerImp = _scroller->scrollerImp();
```
Ditto. No need to specify `<NSScrollerImp>`. Just say `RetainPtr`.

iidmsa replied:
Fixed.

rniwa commented on Source/WebCore/page/scrolling/mac/ScrollerMac.mm:
```diff
@@ -392,46 +398,50 @@ - (void)invalidate
 void ScrollerMac::updatePairScrollerImps()
 {
-    NSScrollerImp *scrollerImp = m_scrollerImp.get();
+    RetainPtr<NSScrollerImp> scrollerImp = m_scrollerImp.get();
```
Ditto. Say `RetainPtr` instead.

rniwa commented on Source/WebCore/page/scrolling/mac/ScrollerMac.mm:
```diff
@@ -392,46 +398,50 @@ - (void)invalidate
 {
-    m_pair.ensureOnMainThreadWithProtectedThis([this](auto&) {
+    RetainPtr<NSScrollerImp> protectedScrollerImp = m_scrollerImp;
+    RetainPtr<NSScrollerImpPair> protectedPair = m_pair->scrollerImpPair();
```
Ditto. `RetainPtr`.

rniwa commented on Source/WebCore/page/scrolling/mac/ScrollerMac.mm:
```diff
@@ -392,46 +398,50 @@ - (void)invalidate
 void ScrollerMac::mouseExitedScrollbar()
 {
-    m_pair.ensureOnMainThreadWithProtectedThis([this](auto&) {
+    RetainPtr<NSScrollerImp> protectedScrollerImp = m_scrollerImp;
```
Ditto.

rniwa commented on Source/WebCore/page/scrolling/mac/ScrollerPairMac.h:
```diff
@@ -48,8 +49,9 @@ class ScrollingTreeScrollingNode;
 
 // Controls a pair of NSScrollerImps via a pair of ScrollerMac. The NSScrollerImps need to remain internal to this class.
-class ScrollerPairMac : public ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr<ScrollerPairMac> {
+class ScrollerPairMac : public ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr<ScrollerPairMac>, public CanMakeCheckedPtr<ScrollerPairMac> {
```
Why are we making this class support CheckedPtr? We can just use RefPtr instead?

iidmsa replied:
Resorted to `ThreadSafeWeakPtr`

rniwa commented on Source/WebCore/page/scrolling/mac/ScrollerMac.h:
```diff
@@ -48,7 +48,7 @@ class ScrollerMac {
     void attach();
 
-    ScrollerPairMac& pair() { return m_pair; }
+    ScrollerPairMac* pair() const { return m_pair.get().get(); }
```
Oh, this is not at all safe. Between the time get() is called and before this value is used, the object can be free'd in another thread. We need to make this function return RefPtr instead.

iidmsa replied:
Fixed.

rniwa commented on Source/WebCore/page/scrolling/mac/ScrollerMac.mm:
```diff
@@ -203,8 +205,10 @@ - (NSAppearance *)effectiveAppearanceForScrollerImp:(NSScrollerImp *)scrollerImp
     if (!_scroller)
         return [NSAppearance currentDrawingAppearance];
+    // FIXME: This is a static analysis false positive.
+    SUPPRESS_UNRETAINED_ARG
```
Please place this on the same line as where NSAppearanceNameDarkAqua and NSAppearanceNameAqua appears.

iidmsa replied:
Done.

iidmsa committed:
Address Safer cpp failures in mac/ScrollerMac
https://bugs.webkit.org/show_bug.cgi?id=291832
rdar://problem/149656606

Reviewed by Ryosuke Niwa.

Fix clang static analyzer warnings in mac/ScrollerMac.

* Source/WebCore/SaferCPPExpectations/NoUncountedMemberCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UnretainedCallArgsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UnretainedLocalVarsCheckerExpectations:
* Source/WebCore/page/scrolling/mac/ScrollerMac.h:
(WebCore::ScrollerMac::pair):
* Source/WebCore/page/scrolling/mac/ScrollerMac.mm:
(-[WebScrollbarPartAnimationMac setCurrentProgress:]):
(-[WebScrollerImpDelegateMac effectiveAppearanceForScrollerImp:]):
(-[WebScrollerImpDelegateMac setUpAlphaAnimation:featureToAnimate:animateAlphaTo:duration:]):
(WebCore::ScrollerMac::attach):
(WebCore::ScrollerMac::updateValues):
(WebCore::ScrollerMac::updateScrollbarStyle):
(WebCore::ScrollerMac::updatePairScrollerImps):
(WebCore::ScrollerMac::mouseEnteredScrollbar):
(WebCore::ScrollerMac::mouseExitedScrollbar):
(WebCore::ScrollerMac::lastKnownMousePositionInScrollbar const):
(WebCore::ScrollerMac::visibilityChanged):
(WebCore::ScrollerMac::updateMinimumKnobLength):
(WebCore::ScrollerMac::scrollbarState const):
* Source/WebCore/page/scrolling/mac/ScrollerPairMac.h:

Canonical link: https://commits.webkit.org/293949@main
